[Brennanzuz] iP

Brennanzuz opened the pull request:
Week 4 commits up till Level-3.

Brennanzuz committed:
Added level-0. greet

Implemented an initial skeletal version of the Duke that simply greets the user and exits.

Brennanzuz committed:
Added greet, echo and exit

Added DukePrinter class to handle printing the output with the correct horizontal line separators and indents.
Added ability to echo all user input, except bye, which would cause the program to terminate.
Moved relevant printing code to DukePrinter.

Brennanzuz committed:
Added Add, List functionality

Added userMessages array to store user messages and echo each element when "list" is entered by the user.
Added printMessages which echoes each element in a given array with the 1-based index of the element.

Brennanzuz committed:
Added "Mark as Done".

Added Task class to store the description of added messages and its marked status.
Added ability to return the string of the task represented with its marked status in brackets in front.
Added commands to mark and unmark with the 2nd argument in the input delimited by spaces to be the 1-based index of the task to be marked.
Replaced messages with tasks.

lipkuangchng commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,49 @@
+        String[] user_input_words = user_input.split(" ");
+        while (!Objects.equals(user_input, "bye")) {
+            switch (user_input_words[0]) {
+            case "list":
```
Avoid magic strings, so perhaps for each of these cases, you can say for ex: LIST = "list" and reference the constant in the if statement

lipkuangchng commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,49 @@
+            default:
+                addTask(new Task(user_input));
+                break;
+            }
```
Good indentation for switch statement and case clauses

leonghuenweng commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,49 @@
+        DukePrinter.printMessage("Nice! I've marked this task as done:\n" + "    " + tasks[taskIndex - 1].getTaskString());
+    }
+    public static void unmarkTask(int taskIndex) {
+        //TODO: ensure taskIndex is not <= 0
```
_I like how you plan your exception handlers ahead before implementing them!_

leonghuenweng commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,38 @@
+        setDone(false);
+    }
+
+    public String getStatusIcon() {
```
Looks good to me! The name "getStatusIcon" is concise and gives me a clear idea its function!

Brennanzuz committed:
Added "Level 4. ToDos, Events, Deadlines"

Added 3 new commands of todo(), deadline(by), and event(from, to) that the user can type into the console.
Added 3 subclasses of Task to allow for the 3 new commands.
Added prefixes 'T', 'D' and 'E' in the returned display of the task to correspond to the type of task being referred to.
Added printAddedTask function to allow easier display of tasks in list
Changed name of project to DukeNUS.
Changed the welcome message logo
Removed the default command to add task when no command is specified.

leonghuenweng commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,49 @@
+        DukePrinter.printWelcomeMessage();
+        String user_input = scanner.nextLine();
+        String[] user_input_words = user_input.split(" ");
+        while (!Objects.equals(user_input, "bye")) {
```
The cases here are handled well! no "Arrowhead style code", good job

leonghuenweng commented on src/main/java/DukePrinter.java:
```diff
@@ -0,0 +1,34 @@
+        + "    | |_| | |_| |   <  __/\n"
+        + "    |____/ \\__,_|_|\\_\\___|\n";
+
+    public static void printHorizontalLine() {
```
perhaps this function can just be named as "printLine"?

leonghuenweng reviewed (commented):
In general everything looks good!

Brennanzuz committed:
Improved code quality.

Added Javadoc comments on almost all functions.
Added Javadoc comments on classes.
Added comments on some variables.
Added error throwing for marking and unmarking tasks that are out of bounds.
Moved printing messages for marking and unmarking to DukeNUSPrinter.java.
Fixed some variable names using snake_case instead of camelCase.

lipkuangchng commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,49 @@
+    public static void unmarkTask(int taskIndex) {
+        //TODO: ensure taskIndex is not <= 0
+        tasks[taskIndex - 1].setDone(false);
+        DukePrinter.printMessage("OK, I've marked this task as not done yet:\n" + "    " + tasks[taskIndex - 1].getTaskString());
```
Perhaps you may want to split the line length to ensure it fits the required number of chars per line

lipkuangchng commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,49 @@
+import java.util.Objects;
+import java.util.Scanner;
+import static java.lang.Integer.parseInt;
```
Good listing of imported classes!

lipkuangchng reviewed (commented):
There are no major coding standard violations. Good job!

Brennanzuz committed:
Added "Level-5: Handle Errors".

Added packages "io" and "tasks" and moved corresponding classes into them.
Added error handling try-catch statements in switch statement for interpreting commands in user input.
Changed all error messages to be more consistent.
Moved main switch statement to `interpret_command()` function.
Fixed javadoc comments in DukeNUSPrinter.java to be neater.
Removed `isTaskIndexValid()` function.

Brennanzuz committed:
Added "Level-6: Handle Errors".

Added new command "delete" that accepts 1 argument for the 1-based index of the task to delete.
Added printer function for the new delete command.
Added comments for the new functions.
Changed tasks array to be a dynamically sized array - ArrayList.
Changed caught exception from ArrayIndexOutOfBoundsException to IndexOutOfBoundsException.

Brennanzuz committed:
Added "Level-7: Save".

Added TasksDataRead to read data stored on a text file and insert it into the tasks ArrayList.
Added TasksDataWrite to write all tasks in tasks ArrayList to a text file that persists between instances.

Brennanzuz committed:
Merge branch 'branch-Level-7'

Brennanzuz committed:
Added "A-Jar" Create a JAR file.

Added functionality to run the program from a jar file.
Added function to store the tasks in tasks ArrayList in data/dukeNUS.txt, that will be retrieved by future instances of the program.
Added another parameter to all Task classes to allow functionality with the TasksDataWrite class and for the saved file to carry information on the done status of the task.

Brennanzuz committed:
Added "Level-8" Dates and Times.

Changed the way commands to create a deadline and event are split into their arguments. Dates are now inputted using a strict yyyy-MM-dd format.
Added error handling for wrong date formats being used to create tasks.

Brennanzuz committed:
Added "Level-9" Find.

Added a new command "find" and a new case for it under Main.
Change "list" command to call listTasks instead of printTasks.
Changed printTasks to only print tasks to be reused by other functions.

Brennanzuz committed:
Merge branch 'branch-Level-9'

Brennanzuz committed:
Added "A-JavaDoc" JavaDoc.

Added JavaDoc comments to almost all functions.
Changed display of date of tasks to be in yyyy-MM-dd to be more consistent with the inputs and the saving.
Changed the delimiters when writing to the file to include more than just a '/'.
Changed the way a saved file is read with the new delimiters.

chinhan99 commented on src/main/java/DukeNUS.java:
```diff
@@ -0,0 +1,205 @@
+                DukeNUSPrinter.printMessage("☹ Error: The task index you have entered is not a number.");
+            } catch (NullPointerException exception) {
+                DukeNUSPrinter.printMessage("☹ Error: The task index you have provided returns a null task.");
+            }
```
Noticed that there are multiple print messages within your try-catch statements. Maybe you could consider having DukeNUSPrinter, or other classes to handle the printing of the error messages!

chinhan99 commented on src/main/java/io/DukeNUSPrinter.java:
```diff
@@ -0,0 +1,117 @@
+     */
+    public static void printAddedTask(String taskDisplayDescription, int totalTaskCount) {
+        printHorizontalLine();
+        System.out.println(INDENT + "Got it. I've added this task:");
+        System.out.println(INDENT + taskDisplayDescription);
```
Since you have the printMessage() method, a similar method could be used instead of using system.out.println.

Brennanzuz committed:
Added "A-MoreOOP" Use More OOP.

Added a new package and class Command to store all functions that edit the tasks ArrayList.
Removed unused packages.

chinhan99 commented on src/main/java/io/TasksDataRead.java:
```diff
@@ -0,0 +1,91 @@
+    private static void interpretSaveEntry(String[] savedEntryWords, ArrayList<Task> savedTasks) {
+        switch (savedEntryWords[0]) {
+        case "todo":
+            String[] todoInput = savedEntryWords[1].split("/marked", 2);
```
Could consider using constants for "/marked" .

chinhan99 commented on src/main/java/DukeNUS.java:
```diff
@@ -0,0 +1,142 @@
+            } catch (ArrayIndexOutOfBoundsException exception) {
+                DukeNUSPrinter.printMessage("☹ Error: The index of the task is missing or out of bounds.");
+            } catch (NumberFormatException exception) {
+                DukeNUSPrinter.printMessage("☹ Error: The task index you have entered is not a number.");
+            } catch (NullPointerException exception) {
+                DukeNUSPrinter.printMessage("☹ Error: The task index you have provided returns a null task.");
+            }
+            break;
+        case "todo":
+            try {
+                Command.addTodo(tasks, new Todo(userInputWords[1], false));
+            } catch (ArrayIndexOutOfBoundsException exception) {
+                DukeNUSPrinter.printMessage("☹ Error: please specify the description of the todo.");
+            }
+            break;
+        case "deadline":
+            try {
+                String[] deadlineInput = userInputWords[1].split("/by", 2);
+                Command.addDeadline(tasks, new Deadline(deadlineInput[0], deadlineInput[1], false));
+            } catch (ArrayIndexOutOfBoundsException exception) {
+                DukeNUSPrinter.printMessage("☹ Error: incorrect syntax. Correct usage: " +
+                        "`deadline [description] /by [deadline]`.");
+            } catch (DateTimeParseException exception) {
+                DukeNUSPrinter.printMessage("☹ Error: invalid date format. Try the yyyy-MM-dd format (eg. 2000-12-31)");
+            }
+            break;
+        case "event":
+            try {
+                String[] eventInput = userInputWords[1].split("/from|/to", 3);
+                Command.addEvent(tasks, new Event(eventInput[0], eventInput[1], eventInput[2], false));
+            } catch (ArrayIndexOutOfBoundsException exception) {
+                DukeNUSPrinter.printMessage("☹ Error: incorrect syntax. Correct usage: " +
+                        "`event [description] /from [start_date] /to [end_date]`.");
+            } catch (DateTimeParseException exception) {
+                DukeNUSPrinter.printMessage("☹ Error: invalid date format. Try the yyyy-MM-dd format (eg. 2000-12-31)");
+            }
+            break;
+        default:
+            DukeNUSPrinter.printMessage("☹ Error: Unknown command. Please try again.");
+            break;
```
This method is too long. Do consider seperating into sub-methods! Dont let a single method handle everything

chinhan99 reviewed (commented):
Do refrain from writing excessively long code by factoring out sub-methods where possible. Do make use of constants too where possible too. 

Other than that, I see consistent comments on most of your methods . 
Good attempt on your IP!

Brennanzuz committed:
Added "A-MoreOOP" Use More OOP.

Added a new package and class Command to store all functions that edit the tasks ArrayList.
Removed unused packages.

Brennanzuz committed:
Added "A-UserGuide" User Guide.

Added the Command.java class from the previous commit after mistakenly not committing it.
Changed the README.md to be related to the project and not the template.

Brennanzuz committed:
Revert "Added "A-MoreOOP" Use More OOP."

This reverts commit 1cfe17e95a05195315f76d090987f1aa88d29bd2.

Brennanzuz committed:
Merge tag 'A-UserGuide'

Brennanzuz committed:
Added "A-UserGuide" User Guide again again.

Copied the contents of README.md to docs\README.md
Fixed a typo

Brennanzuz committed:
Added "A-Release" Release.

Changed the file path to write and read the data txt file on the same folder level as the executable.
